### PR TITLE
TIFF alpha channel

### PR DESCRIPTION
**Pull request by ShirleyNekoDev**
- refactored alpha channel computation and storage
- renamed some PictureExportFormat stuff and replaced `isTransparencySupported` with `transparencyType`
- added TIFF alpha channel export (#1651)
- PR based on top of PR 1659

### CODE REVIEW

**Comment by ShirleyNekoDev** [comment]
Note: The TIFF tag "extra samples" is an extension to the TIFF baseline. Some software might not support it. 
From my experiments even software which can decode alpha samples, might not accept in fp32 format (e.g. IrfanViewdoesn't work; Gimp works). We should implement a UI warning when working on #1650 in another PR.

File: chunky/src/java/se/llbit/chunky/renderer/export/PictureExportFormat.java [hunk 1]
```diff
@@ -50,12 +52,20 @@ default String getDescription() {
...
+   */
+  default AlphaBuffer.Type transparencyType() {
+    return AlphaBuffer.Type.UNSUPPORTED;
+  }
+
+  /**
+   * @return true, if the export formats wants preprocessed buffer data<br>
+   *         false, if the export format uses the unprocessed sampling data
    */
-  default boolean isTransparencySupported() {
```
Comment (leMaik) [inline chunky/src/java/se/llbit/chunky/renderer/export/PictureExportFormat.java:57]: This breaks plugins that adds export formats.

File: chunky/src/java/se/llbit/chunky/renderer/export/PictureExportFormat.java [hunk 2]
```diff
@@ -50,12 +52,20 @@ default String getDescription() {
...
-    return false;
+  default boolean wantsPostprocessing() {
```
Comment (leMaik) [inline chunky/src/java/se/llbit/chunky/renderer/export/PictureExportFormat.java:67]:
```suggestion
  default boolean isPostprocessed() {
```

File: chunky/src/java/se/llbit/chunky/renderer/export/PictureExportFormat.java [hunk 3]
```diff
@@ -50,12 +52,20 @@ default String getDescription() {
   String getExtension();
 
   /**
-   * Check if this format supports transparency (used for transparent sky).
+   * Note: It depends on the scene settings if the alpha buffer will be available on export.
    *
-   * @return True if this format supports transparency, false otherwise
+   * @return the required format for the alpha buffer or {@link AlphaBuffer.Type#UNSUPPORTED} if alpha is not supported.
+   */
+  default AlphaBuffer.Type transparencyType() {
```
Comment (leMaik) [inline chunky/src/java/se/llbit/chunky/renderer/export/PictureExportFormat.java:59]: Rename to keep Java getter style:

```suggestion
  default AlphaBuffer.Type getTransparencyType() {
```

File: chunky/src/java/se/llbit/chunky/renderer/scene/PreviewRayTracer.java [hunk 4]
```diff
@@ -83,7 +83,7 @@ public static double skyOcclusion(Scene scene, WorkerState state) {
 
   /**
    * Find next ray intersection.
-   * @return Next intersection
+   * @return true if intersected, false if no intersection has been found
```
Comment (leMaik) [inline chunky/src/java/se/llbit/chunky/renderer/scene/PreviewRayTracer.java:86]: Nice :+1:

File: chunky/src/java/se/llbit/chunky/renderer/scene/Scene.java [hunk 5]
```diff
@@ -486,9 +485,9 @@ public synchronized void copyState(Scene other, boolean copyChunks) {
       height = other.height;
       backBuffer = other.backBuffer;
       frontBuffer = other.frontBuffer;
-      alphaChannel = other.alphaChannel;
       samples = other.samples;
     }
+    alphaBuffer.reset();
```
Comment (leMaik) [inline chunky/src/java/se/llbit/chunky/renderer/scene/Scene.java:490]: Previously, the alpha buffer of the other scene was copied. Why was this removed?

File: chunky/src/java/se/llbit/chunky/renderer/scene/Scene.java [hunk 6]
```diff
@@ -2631,7 +2563,7 @@ public void setUseCustomWaterColor(boolean value) {
       ((Configurable) postProcessingFilter).storeConfiguration(postprocessJson);
       json.add("postprocessSettings", postprocessJson);
     }
-    json.add("outputMode", outputMode.getName());
+    json.add("outputMode", pictureExportFormat.getName());
```
Comment (leMaik) [inline chunky/src/java/se/llbit/chunky/renderer/scene/Scene.java:2566]: Make sure to keep the old names to not make this a breaking change for existing scene files.

File: chunky/src/java/se/llbit/math/Octree.java [hunk 7]
```diff
@@ -37,8 +37,8 @@
 import se.llbit.chunky.renderer.scene.Scene;
 import se.llbit.chunky.world.Material;
 import se.llbit.log.Log;
-import se.llbit.util.PositionalInputStream;
-import se.llbit.util.PositionalOutputStream;
+import se.llbit.util.io.PositionalInputStream;
```
Comment (leMaik) [inline chunky/src/java/se/llbit/math/Octree.java:40]: Is moving these classes really necessary?

**Review by leMaik — CHANGES_REQUESTED** [review]

File: chunky/src/java/se/llbit/chunky/renderer/export/PictureExportFormat.java (hunk 2 above)
Comment (ShirleyNekoDev) [reply chunky/src/java/se/llbit/chunky/renderer/export/PictureExportFormat.java:76]: This would break plugin compatability

Comment (ShirleyNekoDev) [reply chunky/src/java/se/llbit/chunky/renderer/export/PictureExportFormat.java:76]: Also I like this name because it describes if the format expects a buffer with postprocessed data before exporting. See `Scene::writeFrame()`

File: chunky/src/java/se/llbit/chunky/renderer/scene/Scene.java (hunk 5 above)
Comment (ShirleyNekoDev) [reply chunky/src/java/se/llbit/chunky/renderer/scene/Scene.java:490]: Scene resetting/updating is pretty complicated right now and I'm not sure if the alphaBuffer is cleared every time something changes for the camera perspective. Also when the export format changes, it might have to be recalculated now.
Better save to not keep it here and recalculate this 1 frame before export.
(But if everybody is sure that the buffer can be copied, we can bring it back -- and make the scene-updating easier to understand :)

File: chunky/src/java/se/llbit/math/Octree.java (hunk 7 above)
Comment (ShirleyNekoDev) [reply chunky/src/java/se/llbit/math/Octree.java:40]: I like structure, especially in the util-package :)
Not really necessary, but there are many classes in it which do very different things. And I did not expect plugins to use these IO-classes for now.

File: chunky/src/java/se/llbit/chunky/renderer/scene/Scene.java (hunk 5 above)
Comment (leMaik) [reply chunky/src/java/se/llbit/chunky/renderer/scene/Scene.java:490]:
```suggestion
    alphaBuffer.reset(); // TODO could we copy it without resetting if the export format didn't change?
```

File: chunky/src/java/se/llbit/chunky/renderer/export/PictureExportFormat.java (hunk 2 above)
Comment (leMaik) [reply chunky/src/java/se/llbit/chunky/renderer/export/PictureExportFormat.java:76]: Oh wait, we had `wantsPostprocessing` before? :grimacing: Weird name, but allright.

**Review by leMaik — APPROVED** [review]

**Review by leMaik — APPROVED** [review]

**Comment by ShirleyNekoDev** [comment]
I'll rebase this once #1659 is merged 😵‍💫
